[image-picker][iOS] Migrate the module to the Sweet API

bbarthec opened the pull request:
# Why

[ENG-3078](https://linear.app/expo/issue/ENG-3078/audit-expo-image-picker-ios-rewrite-to-sweet-api)
Superseded https://github.com/expo/expo/pull/15833
Part of converting each and every module to the Sweet API.
I've converted module as-is with no logic change, only splitting and refactoring into logically-related parts.
The code still need some attention to get rid of deprecations that came alongside with iOS 15.

# How

- Renamed pod from EXImagePicker to ExpoImagePicker and prepared the podspec for Swift
- Migrated to the Swift files:
  - `Arguments.swift` for explicit types & structures for JS arguments (autoconversion from JS arguments)
  - `CameraPermissionRequester.swift` and `MediaLibraryPermissionRequesters` for encapsulating permissions-related functionality
  - `Errors.swift` for all errors used in the module
  - `ImagePickerDelegate.swift` for handling all `UIImagePickerController`-related delegates
  - `ImagePickerModule.swift` for describing the module's JS API and providing basic logic (forwarding to the respective parts and storing `imagePickingActionContext` (object storing the current picking action elements, like options or promise)
  - `MediaHandler.swift` for handling the response for the `UIImagePickerController` and shaping the proper response
  - `Response.swift` for elegant response creation. I've used enums to obtain Union Type pattern from Typescript (to be able to operate on single type that encapsulates multiple different ones).
  - `StatusBarVisibilityController.swift` being responsible for hiding/showing the StatusBar when cropping is enabled (addressing known bug in iOS AppKit SDK)
- Created `expo-module.config.json` with iOS-related things

# Test Plan

- [x] Checked the module using dedicated `ncl` screen with no regression
- [x] Checked the `test-suite`
- [ ] Build the bare expo app (used `bare-expo`) and played a bit with the module

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

bbarthec committed:
[modules-core][iOS] Expose logger

Expose EXLogManager to Sweet API to able to log events without promise context

bbarthec committed:
[image-picker][iOS] Create explicit types for method's arguments and JS return values

bbarthec committed:
[image-picker][iOS] Migrate permission requesters to Swift

bbarthec committed:
[image-picker][iOS] Migrate main functions to Swift

I've tried to decompose main class as much as possible.
I propose such components:
- main `ImagePickerModule` to describe JS interface only with basic logic
- permission requesters classes to handle permissions
- `ImagePickerDelegate` to handle all logic related to `UIImagePickerController` (like managing StatusBar visibility in certain cases)
- `MediaHandler` to handle response generated from `UIImagePickerController`/`ImagePickerDelegate` (like constructing the response)

bbarthec committed:
[modules-core][iOS] Stricten nullability rule for path generation method

bbarthec committed:
[image-picker][iOS] Delete Objective-C files and rename podspec

bbarthec committed:
[iOS] Reinstall pods

bbarthec committed:
[image-picker][iOS] Create expo-module.config.json

bbarthec committed:
[iOS][bare-expo] Align SDWebImage version & reinstall pods

bbarthec committed:
[image-picker][iOS] Fix non working options

- Add ImagePickerController delegate to pickingContext object to prevent it's deallocation before picking completion
- Fix reading EXIFs as these are not always present in the MediaMetadata object.
- Fix Obj-C C-style enums interoperability with Swift by passing over raw interger value instead of the enum value itself (C-style enum value is casted to string in Swift)
- Move reading cropping rect to be lazy as it's not present when image is not edited
- Fixed picking videos by passing proper kUTType to the picker controller

lukmccall commented on packages/expo-image-picker/ios/MediaHandler.swift:
```diff
@@ -0,0 +1,275 @@
+    } catch {
+      return completion(.Failure(UnexpectedError(error)))
+    }
+
```
```suggestion
```

lukmccall commented on packages/expo-image-picker/ios/MediaHandler.swift:
```diff
@@ -0,0 +1,275 @@
+          let imageData = try Data.init(contentsOf: fileUrl)
+          return imageData
+        } catch {
+          throw FailedToReadImageDataError(reason: error)
```
I think we should reject here instead of throwing an exception 🤔

lukmccall commented on packages/expo-image-picker/ios/MediaHandler.swift:
```diff
@@ -0,0 +1,275 @@
+        }
+      }
+      if (data == nil) {
+        throw FailedToReadImageDataForBase64Error()
```
same here

lukmccall commented on packages/expo-image-picker/ios/MediaLibraryPermissionRequesters.swift:
```diff
@@ -0,0 +1,102 @@
+    return PHAccessLevel.readWrite
+  }
+}
+
```
```suggestion
```

lukmccall commented on packages/expo-image-picker/ios/ImagePickerDelegate.swift:
```diff
@@ -0,0 +1,74 @@
+    self.statusBarVisibilityController.maybeRestoreStatusBarVisibility()
+    self.onMediaPickingResultHandler.didCancelPicking()
+  }
+  
```
```suggestion
```

lukmccall commented on packages/expo-image-picker/ios/ImagePickerDelegate.swift:
```diff
@@ -0,0 +1,74 @@
+  
+// MARK: UIImagePickerControllerDelegate
+  
+  func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: MediaInfo) {
+    DispatchQueue.main.async {
+      picker.dismiss(animated: true) {
+        self.handlePickedMedia(mediaInfo: info)
+      }
+    }
+  }
+
+  func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
+    DispatchQueue.main.async {
+      picker.dismiss(animated: true) {
+        self.handlePickingCancellation()
+      }
+    }
+  }
```
Was this additional indentation intentional?

bbarthec replied:
I'm not really sure what indentation you're referring to, but `swiftlint` is not complaining about anything indentation related at this point.

lukmccall commented on packages/expo-image-picker/ios/ImagePickerModule.swift:
```diff
@@ -0,0 +1,169 @@
+      ]);
+    }
+
+    function("getCameraPermissionsAsync", { (promise: Promise) in
+      self.appContext?.permissions?.getPermissionUsingRequesterClass(CameraPermissionRequester.self,
+                                                                     resolve: promise.resolver,
+                                                                     reject: promise.legacyRejecter)
+    })
+
+    function("getMediaLibraryPermissionsAsync", { (writeOnly: Bool, promise: Promise) in
+      self.appContext?.permissions?.getPermissionUsingRequesterClass(self.getMediaLibraryPermissionRequester(writeOnly),
+                                                                     resolve: promise.resolver,
+                                                                     reject: promise.legacyRejecter)
+    })
+
+    function("requestCameraPermissionsAsync", { (promise: Promise) in
+      self.appContext?.permissions?.askForPermission(usingRequesterClass: CameraPermissionRequester.self,
+                                                     resolve: promise.resolver,
+                                                     reject: promise.legacyRejecter)
+    })
+
+    function("requestMediaLibraryPermissionsAsync", { (writeOnly: Bool, promise: Promise) in
+      self.appContext?.permissions?.askForPermission(usingRequesterClass: self.getMediaLibraryPermissionRequester(writeOnly),
+                                                     resolve: promise.resolver,
+                                                     reject: promise.legacyRejecter)
```
We should reject if the `permissions` module is not present in those functions.

lukmccall reviewed (commented):
Looks good, just a couple of comments ;)

barthap commented on packages/expo-image-picker/ios/ImagePickerDelegate.swift:
```diff
@@ -0,0 +1,74 @@
+ Protocol that describes scenarios we care about while the user is picking media.
+ */
+protocol OnMediaPickingResultHandler {
+  func didPickedMedia(mediaInfo: MediaInfo) -> Void
+  func didCancelPicking() -> Void
```
Nitpick, but these names are inconsistent - `did picked media` is using past form twice 😉 

```suggestion
  func didPickMedia(mediaInfo: MediaInfo) -> Void
  func didCancelPicking() -> Void
```

tsapeta commented on apps/bare-expo/ios/Podfile.lock:
```diff
@@ -85,15 +85,13 @@ PODS:
   - EXImage (0.2.0):
     - React-Core
-    - SDWebImage (~> 5.0)
+    - SDWebImage (~> 5.12.2)
```
Is it necessary in this PR?

tsapeta commented on apps/bare-expo/ios/Podfile.lock:
```diff
@@ -117,27 +115,27 @@ PODS:
   - Expo (44.0.0-beta.1):
     - ExpoModulesCore
-  - expo-dev-client (0.8.0):
+  - expo-dev-client (0.8.1):
```
Afaik it's already updated on master

tsapeta commented on packages/expo-image-picker/expo-module.config.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "platforms": ["ios"],
```
```suggestion
  "platforms": ["ios", "android"],
```

Otherwise it's not autolinked on Android...

tsapeta commented on packages/expo-image-picker/ios/CameraPermissionRequester.swift:
```diff
@@ -0,0 +1,49 @@
+
+import ExpoModulesCore
+
+public class CameraPermissionRequester : NSObject, EXPermissionsRequester {
```
```suggestion
public class CameraPermissionRequester: NSObject, EXPermissionsRequester {
```

tsapeta commented on packages/expo-image-picker/ios/CameraPermissionRequester.swift:
```diff
@@ -0,0 +1,49 @@
+    var status: EXPermissionStatus
+    let cameraUsageDescription = Bundle.main.object(forInfoDictionaryKey: "NSCameraUsageDescription")
+    let microphoneUsageDescription = Bundle.main.object(forInfoDictionaryKey: "NSMicrophoneUsageDescription")
+    if (cameraUsageDescription == nil || microphoneUsageDescription == nil) {
```
```suggestion
    if cameraUsageDescription == nil || microphoneUsageDescription == nil {
```

tsapeta commented on packages/expo-image-picker/ios/CameraPermissionRequester.swift:
```diff
@@ -0,0 +1,49 @@
+    }
+    
+    return [
+      // TODO: (@bbarthec): had to return status.rawValue, because otherwise this value is casted to string in Obj-C
+      "status": status.rawValue
```
```suggestion
      "status": status.rawValue
```

It's pretty common to refer to `rawValue` so I don't think there is something to do 😉

tsapeta commented on packages/expo-image-picker/ios/CameraPermissionRequester.swift:
```diff
@@ -0,0 +1,49 @@
+    }
+  }
+
+  public func getPermissions() -> [AnyHashable : Any] {
```
```suggestion
  public func getPermissions() -> [AnyHashable: Any] {
```

tsapeta commented on packages/expo-image-picker/ios/Errors.swift:
```diff
@@ -0,0 +1,110 @@
+
+import ExpoModulesCore
+
+internal struct PermissionsModuleNotFoundError: CodedError {
```
Could you migrate these errors to be the exceptions?
```suggestion
internal class PermissionsModuleNotFoundException: Exception {
```

tsapeta commented on packages/expo-image-picker/ios/ImagePickerDelegate.swift:
```diff
@@ -0,0 +1,74 @@
+}
+
+/**
+ This delegate is responsible for responing to any events that are happening in `UIImagePickerController`.
+ It then forward them back in unified way via `ImagePickerResultHandlerDelegate`.
```
```suggestion
 This delegate is responsible for responding to any events that are happening in `UIImagePickerController`.
 It then forwards them back in unified way via `ImagePickerResultHandlerDelegate`.
```

tsapeta commented on packages/expo-image-picker/ios/ImagePickerDelegate.swift:
```diff
@@ -0,0 +1,74 @@
+
+ The functionality of this delegate is separated from the main module class for two reasons:
+ 1) main module cannot inherit from `NSObject` (and that's required by three protocols we must conform to),
+ because it already inhertis from `Module` class and Swift language does not allow multiple inheritance,
```
```suggestion
 because it already inherits from `Module` class and Swift language does not allow multiple inheritance,
```

tsapeta commented on packages/expo-image-picker/ios/ImagePickerDelegate.swift:
```diff
@@ -0,0 +1,74 @@
+  }
+  
+  
+// MARK: UIImagePickerControllerDelegate
```
```suggestion
  // MARK: - UIImagePickerControllerDelegate
```

tsapeta commented on packages/expo-image-picker/ios/ImagePickerDelegate.swift:
```diff
@@ -0,0 +1,74 @@
+    }
+  }
+  
+// MARK: UIAdaptivePresentationControllerDelegate
```
```suggestion
  // MARK: - UIAdaptivePresentationControllerDelegate
```

tsapeta commented on packages/expo-image-picker/ios/ImagePickerDelegate.swift:
```diff
@@ -0,0 +1,74 @@
+    self.handlePickingCancellation()
+  }
+  
+// MARK: UINavigationControllerDelegate
```
```suggestion
  // MARK: - UINavigationControllerDelegate
```

tsapeta commented on packages/expo-image-picker/ios/ImagePickerModule.swift:
```diff
@@ -0,0 +1,169 @@
+import UIKit
+import ExpoModulesCore
+
+typealias MediaInfo = [UIImagePickerController.InfoKey : Any]
```
```suggestion
typealias MediaInfo = [UIImagePickerController.InfoKey: Any]
```

tsapeta commented on packages/expo-image-picker/ios/ImagePickerModule.swift:
```diff
@@ -0,0 +1,169 @@
+  let imagePickerDelegate: ImagePickerDelegate
+}
+
+public class ImagePickerModule : Module, OnMediaPickingResultHandler {
+
```
```suggestion
public class ImagePickerModule: Module, OnMediaPickingResultHandler {
```

tsapeta commented on packages/expo-image-picker/ios/ImagePickerModule.swift:
```diff
@@ -0,0 +1,169 @@
+  private var currentPickingContext: PickingContext?
+
+  private func getMediaLibraryPermissionRequester(_ writeOnly: Bool) -> AnyClass {
+    if (writeOnly) {
```
```suggestion
    if writeOnly {
```

tsapeta reviewed (commented):
Could you please run linter against this? There are a lot of things that it can nitpick to in here.

bbarthec committed:
[image-picker][iOS] Apply suggestions from review

- Address all swiftlint issues
- Moved all permissions requesters to single file
- Prefixed these failenames that were named in a generic way (like `Reposnse`)
- Refactored `Errors` into `Exceptions`
- Cleanup code flow, especially in `MediaHandler`. Split too comlex functions in many simple ones.

bbarthec committed:
[image] Remove accidental dependency bump

bbarthec committed:
[image-picker][iOS] coallesce the exceptions & fix inline docs

lukmccall commented on packages/expo-image-picker/ios/MediaHandler.swift:
```diff
@@ -0,0 +1,347 @@
+    let gps = imageMetadata[kCGImagePropertyGPSDictionary as String] as? [String: Any]
+    if gps != nil {
+      gps!.forEach { key, value in
+        exif["GPD\(key)"] = value
```
```suggestion
        exif["GPS\(key)"] = value
```
I'm not sure here but according to the comment, it should be GPS.

bbarthec replied:
Superb catch! 👏

lukmccall reviewed (approved):
LGTM 🚀
Just a couple of comments.

bbarthec commented on packages/expo-image-picker/ios/ImagePickerModule.swift:
```diff
@@ -0,0 +1,166 @@
+      return promise.reject(CameraUnavailableOnSimulatorException())
+#else
+      picker.sourceType = .camera
+      picker.cameraDevice = context.options.cameraType == .front ? .front : .rear
```
Good catch! 👏 
At this point context is `nil` anyway 😅

bbarthec commented on packages/expo-image-picker/ios/MediaHandler.swift:
```diff
@@ -0,0 +1,347 @@
+                                                           exif: exif))
+        completion(.success(result))
+      }
+    } catch let exception as Exception {
```
I'm not really sure 🤔
`CodedException` is a protocol and `Exception` is implementation that I subclass anyway.
I guess using one or another at this point is rather interchangeable. Generally speaking I expect some `Exception` that I have control over, but Swift is forcing me to catch generic `Error` as well, so this other catch block is unreachable anyway. That's a pity that Swift does not allow annotation around the exception type the function might throw. 😞 

Answering your question: we might establish some convention around catching these `CodedException`/`Exceptions` or let developers decide.

tsapeta replied:
I guess you meant `CodedError`, not `CodedException` 😅 
I'd recommend not to use `CodedError` — I replaced it with exceptions, but I've left it for some time to keep backwards compatibility with the modules that already used `CodedError` and were released already

bbarthec replied:
So I'm leaving it as is 👀

bbarthec commented on packages/expo-image-picker/ios/MediaHandler.swift:
```diff
@@ -0,0 +1,347 @@
+    mediaInfo: MediaInfo,
+    options: ImagePickerOptions
+  ) throws -> (imageData: Data?, fileExtension: String) {
+    let compressionQuality = options.quality ?? DEFAULT_QUALITY
```
Well, we cannot 😞 
Some logic depends on whether it's `nil` or not.
See lines below: https://github.com/expo/expo/pull/15977/files#diff-80fe791d27b3058baa67fb1389774a8e356f4a81f139603d0abcceb2350cb659R160

bbarthec committed:
[iOS][image-picker] Apply review suggestions

tsapeta commented on apps/bare-expo/ios/Podfile.lock:
```diff
@@ -1287,7 +1287,7 @@ SPEC CHECKSUMS:
   boost: a7c83b31436843459a1961bfd74b96033dc77234
   CocoaLumberjack: 543c79c114dadc3b1aba95641d8738b06b05b646
-  DoubleConversion: cf9b38bf0b2d048436d9a82ad2abe1404f11e7de
+  DoubleConversion: 831926d9b8bf8166fd87886c4abab286c2422662
```
I guess something went wrong 🤔 Old React Native version maybe?

tsapeta commented on packages/expo-image-picker/ios/ImagePickerHandler.swift:
```diff
@@ -0,0 +1,72 @@
+  // MARK: - UINavigationControllerDelegate
+
+  func navigationController(_ navigationController: UINavigationController, willShow viewController: UIViewController, animated: Bool) {
+    self.statusBarVisibilityController.maybePreserveVisibilityAndHideStatusBar(self.hideStatusBarWhenPresented)
```
You can omit `self.` when it's not needed 😉 
```suggestion
    statusBarVisibilityController.maybePreserveVisibilityAndHideStatusBar(hideStatusBarWhenPresented)
```

tsapeta commented on packages/expo-image-picker/ios/ImagePickerHandler.swift:
```diff
@@ -0,0 +1,72 @@
+  // MARK: - UIAdaptivePresentationControllerDelegate
+
+  func presentationControllerDidDismiss(_ presentationController: UIPresentationController) {
+    self.handlePickingCancellation()
```
```suggestion
    handlePickingCancellation()
```

bbarthec committed:
[iOS][image-picker] Remove redundant s references wherever it seems too verbose

bbarthec committed:
[iOS][no ci] Apply suggestions

bbarthec committed:
[iOS][no ci] Fix implicit `self` references in closures
